### PR TITLE
Improvements for "Available RAM"

### PR DESCRIPTION
**Pull request by Martinski4GitHub**
Code improvements to obtain a better estimate for the current amount of "Available RAM" to compare with "Required RAM."

### CODE REVIEW

**Comment by Martinski4GitHub** [comment]
@ExtremeFiretop,

As mentioned yesterday, I decided to take another look at the code getting the "Available RAM" value. I was reviewing log entries of memory stats from my router & my cousin's router and have come to the conclusion that, by itself, the "Available RAM" reported by "**/proc/meminfo**" in ASUS routers is not always a reliable metric for available memory. I know it's an estimate based on several factors involved (active vs inactive pages, file-backed vs anonymous pages, current I/O activity buffers, dirty vs writeback page levels, the swap file cache, etc.), and in general, it tends to be extremely conservative in its estimation.

So after going over some C++ code related to memory management that I wrote a while ago for another project at work, I've written a better way to get the estimate of the current amount of "Available RAM" so I submitted this PR with the new approach. Please take a look and, when you get a chance, test the new code on your own routers. 

I've already run testing & verification on my RT-AC86U and my cousin's RT-AX86S, and it's all good. The estimate is much better in situations when "Free RAM" is high and when it's becoming low. Another good thing about this new approach is that it works the same way whether the router has 512MB, 1GB, or 2GB RAM.

File: MerlinAU.sh [hunk 1]
```diff
@@ -3094,9 +3109,9 @@ Please manually update to version $minimum_supported_version or higher to use th
     ##---------------------------------------##
     ## Added by ExtremeFiretop [2023-Dec-09] ##
     ##---------------------------------------##
-    # Get the required space for the firmware download and extraction
-    required_space_kb="$(get_required_space "$release_link")"
-    if ! _HasRouterMoreThan256MBtotalRAM_ && [ "$required_space_kb" -gt 51200 ]
+    # Get the required memory for the firmware download and extraction
```
Comment (ExtremeFiretop) [inline MerlinAU.sh:3112]: Approved

File: MerlinAU.sh [hunk 2]
```diff
@@ -3106,10 +3121,10 @@ Please manually update to version $minimum_supported_version or higher to use th
         fi
     fi
 
-    freeRAM_kb="$(get_free_ram)"
+    freeRAM_kb="$(_GetFreeRAM_KB_)"
```
Comment (ExtremeFiretop) [inline MerlinAU.sh:3124]: Approved all these.

File: MerlinAU.sh [hunk 3]
```diff
@@ -1808,10 +1823,10 @@ check_memory_and_prompt_reboot()
 
         # Check available memory again #
         availableRAM_kb="$(_GetAvailableRAM_KB_)"
-        if [ "$availableRAM_kb" -lt "$required_space_kb" ]
+        if [ "$availableRAM_kb" -lt "$requiredRAM_kb" ]
```
Comment (ExtremeFiretop) [inline MerlinAU.sh:1826]: Same thing, all approved.

File: MerlinAU.sh [hunk 4]
```diff
@@ -1631,74 +1631,89 @@ if false ; then echo "388.5.0" ; return 0 ; fi
...
-## Added by Martinski W. [2023-Dec-15] ##
-##-------------------------------------##
+##----------------------------------------##
+## Modified by Martinski W. [2024-Mar-31] ##
+##----------------------------------------##
 _HasRouterMoreThan256MBtotalRAM_()
 {
    local totalRAM_KB
-   totalRAM_KB="$(cat /proc/meminfo | awk -F ' ' '/^MemTotal: /{print $2}')"
+   totalRAM_KB="$(awk -F ' ' '/^MemTotal:/{print $2}' /proc/meminfo)"
```
Comment (ExtremeFiretop) [inline MerlinAU.sh:1640]: Approved the removal of the pipe to lessen overhead.

File: MerlinAU.sh [hunk 5]
```diff
@@ -1631,74 +1631,89 @@ if false ; then echo "388.5.0" ; return 0 ; fi
...
-   theMemAvailable_KB="$((theMemFree_KB + (thePageCache_KB / 2)))"
+   local theMemAvailable_KB  theMemAvail_KB  theMemCache_KB
+   local theMemFree1_KB  theMemFree2_KB  inactivePgs_KB
+
+   _MaxNumber_() { echo "$(($1 < $2 ? $2 : $1))" ; }
+
+   theMemCache_KB="$(awk -F ' ' '/^Cached:/{print $2}' /proc/meminfo)"
+   theMemFree1_KB="$(awk -F ' ' '/^MemFree:/{print $2}' /proc/meminfo)"
+   theMemAvail_KB="$(awk -F ' ' '/^MemAvailable:/{print $2}' /proc/meminfo)"
+   # Assumes that only ~66% of Page Cache can be reclaimed #
```
Comment (ExtremeFiretop) [inline MerlinAU.sh:1677]: I prefer the higher accuracy when comparing to conservatism.
Nice work, I like the fallback for the theMemAvailable_KB variable if MemAvailable is not available.

File: MerlinAU.sh [hunk 6]
```diff
@@ -1631,74 +1631,89 @@ if false ; then echo "388.5.0" ; return 0 ; fi
...
 }
 
-get_free_ram() {
-    # Using awk to sum up the 'free', 'buffers', and 'cached' columns.
-    free | awk '/^Mem:/{print $4}'  # This will return the available memory in kilobytes.
-    ##FOR DEBUG ONLY##echo 1000
+##----------------------------------------##
+## Modified by Martinski W. [2024-Mar-31] ##
+##----------------------------------------##
+_GetFreeRAM_KB_()
```
Comment (ExtremeFiretop) [inline MerlinAU.sh:1692]: Approved.

File: MerlinAU.sh [hunk 7]
```diff
@@ -1631,74 +1631,89 @@ if false ; then echo "388.5.0" ; return 0 ; fi
...
-get_required_space() {
+##----------------------------------------##
+## Modified by Martinski W. [2024-Mar-31] ##
+##----------------------------------------##
+_GetRequiredRAM_KB_()
+{
     local url="$1"
-    local zip_file_size_kb extracted_file_size_buffer_kb
-    local overhead_percentage=50  # Overhead percentage (e.g., 50%)
+    local zip_file_size_bytes  zip_file_size_kb  overhead_kb
```
Comment (ExtremeFiretop) [inline MerlinAU.sh:1704]: Approved cleanup

**Review by ExtremeFiretop — APPROVED** [review]

**Comment by ExtremeFiretop** [comment]
@Martinski4GitHub 

Great work, I like the improvements, as you mentioned here:

>and in general, it tends to be extremely conservative in its estimation.

I generally agree that based on my testing it tended to be more conservative than accurate, so the fact that this improves accuracy, plus works the same way whether the router has 512MB, 1GB, or 2GB RAM, is a major plus.

Merged! Will test it tomorrow it's starting to get late but seems solid enough to merge into dev :)
Will report back more tomorrow!

**Comment by Martinski4GitHub** [comment]
> 
> I generally agree that based on my testing it tended to be more conservative than accurate, so the fact that this improves accuracy, plus works the same way whether the router has 512MB, 1GB, or 2GB RAM, is a major plus.
> 
> Merged! Will test it tomorrow it's starting to get late but seems solid enough to merge into dev :) Will report back more tomorrow!
>
Sounds good. I'm going to bed early tonight as well since I have an early morning meeting tomorrow.

Good night, buddy!!

**Comment by ExtremeFiretop** [comment]
@Martinski4GitHub 

Find the attached results comparing the 2 runs between the old memory method and the new memory method:

**OLD MEMORY METHOD:** 
[OLD_MEM_GT-AXE11000_FW_Update_2024-04-03_22_02_04.log](https://github.com/ExtremeFiretop/MerlinAutoUpdate-Router/files/14860444/OLD_MEM_GT-AXE11000_FW_Update_2024-04-03_22_02_04.log)

**NEW MEMORY METHOD:**
[NEW_MEM_GT-AXE11000_FW_Update_2024-04-03_21_56_25.log](https://github.com/ExtremeFiretop/MerlinAutoUpdate-Router/files/14860445/NEW_MEM_GT-AXE11000_FW_Update_2024-04-03_21_56_25.log)

**Comment by ExtremeFiretop** [comment]
At the last stage of before the flash on the new method it stated:

Required RAM: 227272 KB - **RAM Free: 210568 KB** - **RAM Available: 449770 KB**

Old method stated:

Required RAM: 227272 KB - **RAM Free: 210956 KB** - **RAM Available: 352788 KB**

**Comment by ExtremeFiretop** [comment]
I rather accuracy than conservatism and this seems to accomplish that!

Nice work :) This may even explain why my router looked like it wasn't decreasing in memory usage in my previous tests, when obviously it had to be (since I was loading it with more and more junk.)

Makes me think I should re-run those tests!

**Comment by ExtremeFiretop** [comment]
Oh! And to the point of the user that had low memory, this would technically get him over the minimum requirement I believe as well.

**Comment by ExtremeFiretop** [comment]
FYI regarding Gnuton:

Screenshot_2024-04-03-23-33-36-43_572064f74bd5f9fa804b05334aa4f912

**Comment by ExtremeFiretop** [comment]
And here's the output of the webs_update.sh on Gnuton: https://pastebin.com/886EQVzN#google_vignette

**Comment by Martinski4GitHub** [comment]
@ExtremeFiretop,

Apologies for not responding earlier. My wife's Mom has a bad fall at her home late afternoon on Tuesday. At first, it seemed like it was just some bruises & scrapes, but after the pain got worse overnight, they went to the doctor & found out she had a hairline fracture on both her wrist & ulna. So my wife & I have been driving to my in-law's home in the evenings to check on her and take food to have dinner with them & some fruits, vegetables and easy-to-prepare food so they can eat during the day. Anyway, needless to say, my last few evenings have been occupied so I have done much WRT MerlinAU.

**Comment by ExtremeFiretop** [comment]
> @ExtremeFiretop,
> 
> Apologies for not responding earlier. My wife's Mom has a bad fall at her home late afternoon on Tuesday. At first, it seemed like it was just some bruises & scrapes, but after the pain got worse overnight, they went to the doctor & found out she had a hairline fracture on both her wrist & ulna. So my wife & I have been driving to my in-law's home in the evenings to check on her and take food to have dinner with them & some fruits, vegetables and easy-to-prepare food so they can eat during the day. Anyway, needless to say, my last few evenings have been occupied so I have done much WRT MerlinAU.

No worries I totally understand, I always assume something is just keeping you busy, it's unfortunate to hear it's due to an unfortunate accident. I do hope she recovers quickly and best wishes for you and the family!

I've slowly started poking at Gnuton stuff and tested this and that's basically the TLDR.

**Comment by ExtremeFiretop** [comment]
I'm sure she's appreciated having you visit in the evening :)

**Comment by Martinski4GitHub** [comment]
> 
> No worries I totally understand, I always assume something is just keeping you busy, it's unfortunate to hear it's due to an unfortunate accident. I do hope she recovers quickly and best wishes for you and the family!
> 
Thanks. This evening she seemed much better, so my wife was relieved & also feeling much better about the situation. Her dad was also less anxious, so overall we were all breathing a lot easier this evening.

>
> I'm sure she's appreciated having you visit in the evening :)
>
My in-laws are good folks and have always been willing & ready to help us with our kids since they were babies. Along with my own parents, they took a fair share of babysitting duties, taking them to the doctor when needed, picking them up from school, staying with them at home when they were sick, etc. As a parent, when someone shows your kids how much they love them & demonstrates how much they care about them, you never forget it, so I'm always appreciative of that. Also, despite our disagreements about other topics (e.g. politics, societal issues, etc.), my in-laws have been respectful to us about how my wife and I raised our kids and about our marriage. The proverb: "It takes a village to raise a child" is certainly true, and the "first village" starts with your close relatives.

**Comment by Martinski4GitHub** [comment]
> At the last stage of before the flash on the new method it stated:
> 
> Required RAM: 227272 KB - **RAM Free: 210568 KB** - **RAM Available: 449770 KB**
> 
> Old method stated:
> 
> Required RAM: 227272 KB - **RAM Free: 210956 KB** - **RAM Available: 352788 KB**
>
This looks very good and is comparable to the test results I got when verifying the code with **RT-AC86U** & **RT-AX86S** (both with 512MB RAM).

**Comment by Martinski4GitHub** [comment]
> Oh! And to the point of the user that had low memory, this would technically get him over the minimum requirement I believe as well.
>
Yes, I do think this new code should address the issue with low "Free RAM" but enough "Available RAM" to continue with F/W flashing.

**Pull request merged**